File: Unity/ViewLink/Messages.cs
using System.Collections.Generic;
using UnityEngine;

namespace BlenderBridge.ViewLink
{
    struct ViewUpdated : Comms.IMessage
    {
        public byte Type => (byte)ClientMessageType.ViewUpdated;

        public string ID;
        public int Width;
        public int Height;
        public bool IsPerspective;
        public List<float> ViewMatrix;
        public List<float> WindowMatrix;

        public Matrix4x4 ConvertedViewMatrix => ConvertListToMatrix(ViewMatrix);
        public Matrix4x4 ConvertedWindowMatrix => ConvertListToMatrix(WindowMatrix);

        Matrix4x4 ConvertListToMatrix(IList<float> data)
        {
            if (data.Count != 16)
            {
                Debug.LogError($"Invalid matrix received for View with ID {ID}");

                return Matrix4x4.identity;
            }

            return new Matrix4x4(
                new Vector4(data[0], data[4], data[8], data[12]),
                new Vector4(data[1], data[5], data[9], data[13]),
                new Vector4(data[2], data[6], data[10], data[14]),
                new Vector4(data[3], data[7], data[11], data[15])
            );
        }
    }

    struct ViewDestroyed : Comms.IMessage
    {
        public byte Type => (byte)ClientMessageType.ViewDestroyed;

        public string ID;
    }
    
    struct ObjectVisibility : Comms.IMessage
    {
        public byte Type => (byte)ClientMessageType.ObjectVisibility;

        public string Name;
        public bool Visible;
        public bool Obsolete;
    }
}

File: Blender~/view_link/messages.py
from ..comms.message import IMessage
from ..messages import ServerMessageType, ClientMessageType


class ViewUpdated(IMessage):

    @staticmethod
    def get_type():
        return ClientMessageType.ViewUpdated

    def __init__(self):
        self.view = None
        self.str_id = None

    def initialize(self, view, str_id):
        self.view = view
        self.str_id = str_id
        return self

    def serialize(self):
        data = {}
        data['ID'] = self.str_id
        data['Width'] = self.view.width
        data['Height'] = self.view.height
        data['IsPerspective'] = self.view.is_perspective
        data['ViewMatrix'] = [j for sub in self.view.view_matrix for j in sub]
        data['WindowMatrix'] = [j for sub in self.view.window_matrix for j in sub]
        return data

    def post_deserialize(self):
        pass


class ViewDestroyed(IMessage):

    @staticmethod
    def get_type():
        return ClientMessageType.ViewDestroyed

    def __init__(self):
        self.ID = None

    def initialize(self, id):
        self.ID = id
        return self

    def serialize(self):
        data = self.__dict__
        return data

    def post_deserialize(self):
        pass


class ObjectVisibility(IMessage):

    @staticmethod
    def get_type():
        return ClientMessageType.ObjectVisibility

    def __init__(self):
        self.Name = None
        self.Visible = None
        self.Obsolete = None

    def initialize(self, name, visible, obsolete):
        self.Name = name
        self.Visible = visible
        self.Obsolete = obsolete
        return self

    def serialize(self):
        data = self.__dict__
        return data

    def post_deserialize(self):
        pass
